hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/DefaultCertificateClient.java

```diff
@@ -340,20 +349,52 @@ public synchronized List<X509Certificate> getTrustChain() {
       }
     } else {
       // case before certificate bundle is supported
-      chain.add(getCertificate());
-      X509Certificate cert = getCACertificate();
-      if (cert != null) {
-        chain.add(getCACertificate());
-      }
-      cert = getRootCACertificate();
-      if (cert != null) {
-        chain.add(cert);
+      X509Certificate lastInsertedCert = getCertificate();
+      chain.add(lastInsertedCert);
+      List<X509Certificate> caCertList = getCaCertList();
+      while (!isRootCa(lastInsertedCert)) {
+        Optional<X509Certificate> issuerOpt =
+            getIssuerForCert(lastInsertedCert, caCertList);
+        if (issuerOpt.isPresent()) {
+          X509Certificate issuer = issuerOpt.get();
+          chain.add(issuer);
+          lastInsertedCert = issuer;
+        } else {
+          throw new CertificateException("No issuer found for certificate: " +
+              lastInsertedCert);
+        }
       }
+      //add self-signed certificate to the chain
+      chain.add(lastInsertedCert);
```

Comment:
This lastInsertedCert is added in Line 353 too.

Reply:
The lastInsertedCert is changed during the iteration. The method here does the following: takes a certificate lastInsertedCert, checks if there is an issuer for that certificate, in that case inserts the cert to the list and continues the same thing with the issuer. We end the iteration when there is no issuer left for the lastInsertedCert. This means that we have reached a certificate whose signer is not in our list. But we still haven't added this certificate to our list, which is inserted at the end.

Reply:
I see.